x-pack/legacy/plugins/siem/public/components/embeddables/map_tool_tip/map_tool_tip.tsx

```diff
@@ -0,0 +1,148 @@
+import {
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiLoadingSpinner,
+  EuiOutsideClickDetector,
+} from '@elastic/eui';
+import { FeatureGeometry, FeatureProperty, MapToolTipProps } from '../types';
+import { DraggablePortalContext } from '../../drag_and_drop/draggable_wrapper';
+import { ToolTipFooter } from './tooltip_footer';
+import { LineToolTipContent } from './line_tool_tip_content';
+import { PointToolTipContent } from './point_tool_tip_content';
+import { Loader } from '../../loader';
+
+export const MapToolTip = React.memo<MapToolTipProps>(
+  ({
+    addFilters,
+    closeTooltip,
+    features = [],
+    isLocked,
+    getLayerName,
+    loadFeatureProperties,
+    loadFeatureGeometry,
+  }) => {
+    const [isLoading, setIsLoading] = useState<boolean>(true);
+    const [isLoadingNextFeature, setIsLoadingNextFeature] = useState<boolean>(false);
+    const [featureIndex, setFeatureIndex] = useState<number>(0);
+    const [featureProps, setFeatureProps] = useState<FeatureProperty[]>([]);
+    const [featurePropsFilters, setFeaturePropsFilters] = useState<Record<string, object>>({});
+    const [featureGeometry, setFeatureGeometry] = useState<FeatureGeometry | null>(null);
+    const [, setLayerName] = useState<string>('');
+
+    useEffect(() => {
+      // Early return if component doesn't yet have props -- result of mounting in portal before actual rendering
+      if (
+        features.length === 0 ||
+        getLayerName == null ||
+        loadFeatureProperties == null ||
+        loadFeatureGeometry == null
+      ) {
+        return;
+      }
+
+      // Separate loaders for initial load vs loading next feature to keep tooltip from drastically resizing
+      if (!isLoadingNextFeature) {
+        setIsLoading(true);
+      }
+
+      const fetchFeatureProps = async () => {
+        if (features[featureIndex] != null) {
+          const layerId = features[featureIndex].layerId;
+          const featureId = features[featureIndex].id;
+          const [featureProperties, featureGeo, layerNameString] = await Promise.all([
+            loadFeatureProperties({ layerId, featureId }),
+            loadFeatureGeometry({ layerId, featureId }),
+            getLayerName(layerId),
+          ]);
+
+          // Fetch ES filters in advance while loader is present to prevent lag when user clicks to add filter
+          const featurePropsPromises = await Promise.all(
+            featureProperties.map(property => property.getESFilters())
+          );
+          const featurePropsESFilters = featureProperties.reduce(
+            (acc, property, index) => ({
+              ...acc,
+              [property._propertyKey]: featurePropsPromises[index],
+            }),
+            {}
+          );
+
+          setFeatureProps(featureProperties);
+          setFeaturePropsFilters(featurePropsESFilters);
+          setFeatureGeometry(featureGeo);
+          setLayerName(layerNameString);
+          setIsLoading(false);
```

Comment:
Will this cleanup code (resetting loading indicators) execute if any of the promises in

```
const featurePropsPromises = await Promise.all(...
```

reject?

Reply:
So just dug a bit into this and it would've resulted in an unhandled Promise rejection and the cleanup not running. One option would be to reject, but would need to do that for each promise, so I opted for try/catch/finally and adding an error state. Thanks for calling attention to this!